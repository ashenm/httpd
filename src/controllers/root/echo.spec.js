const echo = require('./echo');
const dispatch = require('../../utilities/status');

jest.mock('../../utilities/status');

describe('root/echo', () => {
  const response = {
    send: jest.fn().mockReturnThis(),
    status: jest.fn().mockReturnThis(),
    type: jest.fn().mockReturnThis()
  };

  const request = {
    headers: { 'X-Key': 'x-test-value' },
    get: jest.fn(),
    is: jest.fn()
  };

  describe('application/json', () => {
    // given
    beforeAll(() => {
      request.get.mockReturnValueOnce(null);
      request.is.mockReturnValueOnce(null);
      jest.spyOn(dispatch, 'success').mockReturnValueOnce('test-dispatch-success-value');
    });

    // when
    beforeAll(() => {
      echo(request, response);
    });

    // then
    it('Should attempt to resolve default response body with valid arity', () => {
      expect(jest.spyOn(dispatch, 'success').mock.calls).toMatchSnapshot();
    });

    it('Should dispatch response status with valid arity', () => {
      expect(response.status.mock.calls).toMatchSnapshot();
    });

    it('Should dispatch response body with valid arity', () => {
      expect(response.send.mock.calls).toMatchSnapshot();
    });

    afterAll(jest.clearAllMocks);
  });

  describe('text/plain', () => {
    // given
    beforeAll(() => {
      request.get.mockReturnValueOnce('text/plain');
      request.is.mockReturnValueOnce(false);
    });

    // when
    beforeAll(() => {
      echo({ ...request, body: 'test-request-body' }, response);
    });

    // then
    it('Should not attempt to resolve default response body', () => {
      expect(jest.spyOn(dispatch, 'success')).not.toBeCalled();
    });

    it('Should dispatch response status with valid arity', () => {
      expect(response.status.mock.calls).toMatchSnapshot();
    });

    it('Should dispatch response body with valid arity', () => {
      expect(response.send.mock.calls).toMatchSnapshot();
    });

    afterAll(jest.clearAllMocks);
  });
});
