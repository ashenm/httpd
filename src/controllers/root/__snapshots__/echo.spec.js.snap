// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`root/echo application/json Should attempt to resolve default response body with valid arity 1`] = `
[
  [
    200,
  ],
]
`;

exports[`root/echo application/json Should dispatch response body with valid arity 1`] = `
[
  [
    "test-dispatch-success-value",
  ],
]
`;

exports[`root/echo application/json Should dispatch response status with valid arity 1`] = `
[
  [
    200,
  ],
]
`;

exports[`root/echo text/plain Should dispatch response body with valid arity 1`] = `
[
  [
    "test-request-body",
  ],
]
`;

exports[`root/echo text/plain Should dispatch response status with valid arity 1`] = `
[
  [
    200,
  ],
]
`;
